File: ex1/ex1-5.c
#include <stdio.h>

int infinity();
void test(int, int);

int main(){
    // 無限ループ C言語も適用順評価
    test(0, infinity());
    return 0;
}

void test(int x, int y){
    if(x == 0){
        return;
    }else{
        return;
    }
}

int infinity(){
    while(1);
    return 1;
}

File: ex1/ex1-5.hs
infinity :: Integer -> Integer
infinity 0 = 1
infinity x = infinity(x)

test :: Integer -> Integer -> Integer
test 0 _ = 1
test _ y = y

-- $ ghci
-- > :load ex1-5.hs
-- > infinity 1 // 無限ループ
-- > test 0 (infinity 1) // y つまり infinity 1 は評価されないので無限ループしない
-- > test 1 (infinity 1) // y つまり infinity 1 が評価されて無限ループする
